File: solutions/aoc2021_day03.js
var fs = require('fs');

function power(pi) {
    var arr = pi.split('\n').map(String);
    var s1 = "";
    var s2 = "";
    for(var j = 0; j < arr[0].length; j++) {
        var c = {'0' : 0, '1': 0};
        for(var i = 0; i < arr.length; i++) {
            var e = arr[i][j];
            c[e] += 1;
        }
        if (c['0'] > c['1']) {
            s1 += "0";
            s2 += "1";
        }
        else {
            s1 += "1";
            s2 += "0";
        }
    }
    var x = parseInt(s1, 2);
    var y = parseInt(s2, 2);
    return x * y;
}

function oxygen_scrubber(arr, type) {
    var arr = pi.split('\n').map(String);
    var rows = Array(arr.length).fill().map((_, i) => i);
    for(var j = 0; j < arr[0].length; j++) {
        var c = {'0' : 0, '1': 0};
        for(var i = 0; i < arr.length; i++) {
            if (rows.includes(i)) {
                var e = arr[i][j];
                c[e] += 1;
            }
        }
        var t = (c['0'] > c['1']) ? '1' : '0';
        if (type === "scrubber") {
            t = (t === '1') ? '0' : '1';
        }
        for (var i = 0; i < arr.length; i++) {
            if (!rows.includes(i)) {
                continue;
            }
            if (arr[i][j] === t) {
                var index = rows.indexOf(i);
                rows.splice(index, 1);
            }
            if (rows.length == 1) {
                var i = rows[0];
                var s = arr[i];
                return parseInt(s, 2);
            }
        }
    }
    return null;
}

function part1(pi) {
    var ans = power(pi);
    console.log(`part1: ${ans}`);
}

function part2(pi) {
    var ans = oxygen_scrubber(pi, "oxygen") * oxygen_scrubber(pi, "scrubber");
    console.log(`part2: ${ans}`);
}

console.log("day03");
var f = fs.readFileSync('aoc2021_day03.txt', 'utf8');
var pi = f.toString();
part1(pi);
part2(pi);

File: solutions/aoc2021_day03.py
def power(arr):
    s1 = ""
    s2 = ""
    for j in range(len(arr[0])):
        c = {'0': 0, '1': 0}
        for i in range(len(arr)):
            c[arr[i][j]] += 1
        if c['0'] > c['1']:
            s1 += "0"
            s2 += "1"
        else:
            s1 += "1"
            s2 += "0"
    x = int(s1, 2)
    y = int(s2, 2)
    ans = x * y
    return ans

def oxygen_scrubber(arr, type):
    rows = set([i for i in range(len(arr))])
    for j in range(len(arr[0])):
        c = {'0': 0, '1': 0}
        for i in range(len(arr)):
            if i in rows:
                c[arr[i][j]] += 1
        if type == "oxygen":
            t = '1' if c['0'] > c['1'] else '0'
        else: # scrubber
            t = '0' if c['0'] > c['1'] else '1'
        for i in range(len(arr)):
            if i not in rows:
                continue
            if arr[i][j] == t:
                rows.remove(i)
            if len(rows) == 1:
                scrubber = int(arr[list(rows)[0]], 2)
                return scrubber
    return None

def part1(arr):
    ans = power(arr)
    print(ans)

def part2(arr):
    x = oxygen_scrubber(arr, "oxygen")
    y = oxygen_scrubber(arr, "scrubber")
    ans = x * y
    print(ans)

with open("aoc2021_day03.txt") as f:
    data = f.read()
    arr = [line for line in data.splitlines()]

print("day03")
part1(arr)
part2(arr)
